public_data_report/hardware_report/hardware_report.py

```diff
@@ -0,0 +1,478 @@
+import boto3
+import click
+from datetime import datetime, timedelta
+import json
+import logging
+import requests
+
+from google.cloud import bigquery
+
+from pyspark.sql import SparkSession, Row
+
+
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def load_data(spark, date_from, date_to):
+    """Load a set of aggregated metrics for the provided timeframe.
+  
+    Returns Spark dataframe containing preaggregated user counts per various dimensions.
+
+    Args:
+        date_from: Start date (inclusive)
+        date_to: End date (exclusive)
+    """
+    bq = bigquery.Client()
+
+    query = """
+  WITH
+    rank_per_client AS (
+      SELECT
+        *,
+        ROW_NUMBER() OVER (PARTITION BY client_id ORDER BY submission_timestamp DESC) AS rn
+      FROM
+        `moz-fx-data-shared-prod.telemetry_stable.main_v4`
+      WHERE
+        DATE(submission_timestamp) >= @date_from
+        AND DATE(submission_timestamp) < @date_to
+    ),
+    latest_per_client_all AS (
+      SELECT
+        *
+      FROM
+        rank_per_client
+      WHERE
+        rn=1
+    ),
+    latest_per_client AS (
+      SELECT
+        environment.build.architecture AS browser_arch,
+        COALESCE(environment.system.os.name,
+            'Other') AS os_name,
+        COALESCE(
+            IF (environment.system.os.name='Linux',
+                REGEXP_EXTRACT(environment.system.os.version, r"^[0-9]+\.[0-9]+"),
```

Comment:
This is grouping minor Linux versions together.

Such grouping is also handled by Transposer (old dashboard had all Linux versions collapsed to a single label). I added it here because now we're starting to produce a BQ table. I feel like it would be good to have data in this table as similar as possible to what we show on the dashboard - then eventually we'll be able to easily make it public and potentially get rid of intermediate S3 storage.

Reply:
Agreed, it is nice to move some of those changes to this table.